File: typescript/src/sprites/sprite.ts
import P5 from 'p5';
import Colors from '../ui/colors';

export default class Sprite {
    p5: P5;
    position: P5.Vector;
    velocity = new P5.Vector();
    diameter: number;
    rotation = 0;
    rotationStep = 0;

    constructor(
        p5: P5,
        position: P5.Vector,
        diameter: number,
        velocity: P5.Vector,
        rotationStep: number = 0
    ) {
        this.p5 = p5;
        this.position = position;
        this.diameter = diameter;
        this.velocity = velocity;
        this.rotationStep = rotationStep;
    }

    update(...args: any): boolean {
        this.position.add(this.velocity);
        this.checkWindowEdges();

        this.rotation += this.rotationStep;

        return true;
    }

    draw(...args: any) {
        this.p5.push();

        this.p5.translate(this.position.x, this.position.y);
        this.p5.rotate(this.rotation);

        this.p5.fill(Colors.EDGE);
        this.p5.noStroke();

        this.p5.ellipse(0, 0, this.diameter);

        this.p5.fill(Colors.DARK);
        this.p5.rectMode(this.p5.CENTER);
        this.p5.rect(0, 0, this.diameter / 4, this.diameter / 4);

        this.p5.pop();
    }

    checkWindowEdges(): boolean {
        const radius = this.diameter / 2;
        const width = this.p5.width;
        const height = this.p5.height;
        const x = this.position.x;
        const y = this.position.y;

        if (x > width + radius) {
            this.position.x = -radius;
            return true;
        } else if (x < -radius) {
            this.position.x = width + radius;
            return true;
        }

        if (y > height + radius) {
            this.position.y = -radius;
            return true;
        } else if (y < -radius) {
            this.position.y = height + radius;
            return true;
        }

        return false;
    }

    collideWith(sprite: Sprite): boolean {
        const distance = this.p5.dist(
            this.position.x,
            this.position.y,
            sprite.position.x,
            sprite.position.y
        );

        if (distance < this.diameter / 2 + sprite.diameter / 2) {
            return true;
        }

        return false;
    }
}

File: rust/src/sprites/sprite.rs
extern crate web_sys;

use core::f64::consts::PI;

use web_sys::CanvasRenderingContext2d;

use crate::colors::Colors;
use crate::vector::Vector;

pub trait Spritable {
    fn update(&mut self);
    fn draw(&self, canvas: CanvasRenderingContext2d);
    fn collide_with(&self, sprite: Sprite) -> bool;
}

#[derive(Copy, Clone)]
pub struct SpriteData {
    pub position: Vector,
    pub velocity: Vector,
    pub diameter: f64,
    pub rotation: f64,
    pub rotation_step: f64,
}

#[derive(Copy, Clone)]
pub struct CanvasDimension {
    pub width: f64,
    pub height: f64,
}

#[derive(Copy)]
pub struct Sprite {
    pub data: SpriteData,
    pub canvas: CanvasDimension,
    pub is_offscreen: bool,
}

impl Clone for Sprite {
    fn clone(&self) -> Self {
        Sprite {
            data: self.data,
            canvas: self.canvas,
            is_offscreen: self.is_offscreen,
        }
    }
}

impl Sprite {
    pub fn new(sprite_data: SpriteData, canvas: CanvasDimension) -> Sprite {
        Sprite {
            data: sprite_data,
            canvas,
            is_offscreen: false,
        }
    }

    fn check_window_edges(&mut self) -> bool {
        let radius = self.data.diameter / 2.0;
        let x = self.data.position.x;
        let y = self.data.position.y;
        let mut result = false;

        if x > self.canvas.width + radius {
            self.data.position.x = -radius;
            result = true;
        } else if x < -radius {
            self.data.position.x = self.canvas.width + radius;
            result = true;
        }

        if y > self.canvas.height + radius {
            self.data.position.y = -radius;
            result = true;
        } else if y < -radius {
            self.data.position.y = self.canvas.height + radius;
            result = true;
        }

        self.is_offscreen = result;

        result
    }
}

impl Spritable for Sprite {
    fn update(&mut self) {
        self.data.position += self.data.velocity;
        self.data.rotation += self.data.rotation_step;

        self.check_window_edges();
    }

    fn draw(&self, canvas: CanvasRenderingContext2d) {
        canvas.save();

        let _result = canvas.translate(self.data.position.x, self.data.position.y);
        let _result = canvas.rotate(self.data.rotation);

        canvas.begin_path();

        let _result = canvas.arc(0.0, 0.0, 40.0, 0.0, 2.0 * PI);

        canvas.fill();

        canvas.set_fill_style(&Colors::Dark.value().into());

        let _result = canvas.translate(-10.0, -10.0);

        canvas.fill_rect(0.0, 0.0, 20.0, 20.0);

        canvas.restore();
    }

    fn collide_with(&self, sprite: Sprite) -> bool {
        let distance = self.data.position.distance(sprite.data.position);

        if distance < self.data.diameter / 2.0 + sprite.data.diameter / 2.0 {
            return true;
        }

        false
    }
}
